core/src/main/scala/kanaloa/reactive/dispatcher/queue/AutoScaling.scala

```diff
@@ -1,138 +1,125 @@
+    case PartialUtilization(utilization) ⇒
+      if (highestUtilization < utilization)
+        context become underUtilized(utilization, start)
+    case s: Sample ⇒
+      context become fullyUtilized(s.poolSize)
+      self ! s
+    case OptimizeOrExplore ⇒
+      if (start.isBefore(Time.now.minus(downsizeAfterUnderUtilization)))
+        processor ! ScaleTo((highestUtilization * downsizeRatio).toInt, Some("downsizing"))
+    case qs: QueryStatus ⇒
+      qs.reply(AutoScalingStatus(partialUtilization = Some(highestUtilization), partialUtilizationStart = Some(start)))
+  }
 
-    val currentSize: PoolSize = workerPool.size
+  private def fullyUtilized(currentSize: PoolSize): Receive = watchingQueueAndProcessor orElse {
+    case Sample(workDone, start, end, poolSize) ⇒
 
-    // Send metrics
-    metricsCollector.send(Metric.PoolSize(currentSize))
-    metricsCollector.send(Metric.PoolUtilized(utilization))
+      val speed: Double = workDone.toDouble / start.until(end).toMillis.toDouble
+      val toUpdate = perfLog.get(poolSize).fold(speed) { oldSpeed ⇒
+        oldSpeed * (1d - weightOfLatestMetric) + (speed * weightOfLatestMetric)
+      }
+      perfLog += (poolSize → toUpdate)
+      context become fullyUtilized(poolSize)
 
-    underUtilizationStreak = if (dispatchWaitWhenFullyUtilized.isEmpty)
-      underUtilizationStreak.map(s ⇒ s.copy(highestUtilization = Math.max(s.highestUtilization, utilization))) orElse Some(UnderUtilizationStreak(LocalDateTime.now, utilization))
-    else None
+    case PartialUtilization(u) ⇒
+      context become underUtilized(u)
 
-    if (underUtilizationStreak.isDefined)
-      downsize
-    else {
-      if (dispatchWaitWhenFullyUtilized.isDefined) {
-        val toUpdate = perfLog.get(currentSize).fold(dispatchWaitWhenFullyUtilized.get) { oldSpeed ⇒
-          val nanos = (oldSpeed.toNanos * (1d - weightOfLatestMetric)) + (dispatchWaitWhenFullyUtilized.get.toNanos * weightOfLatestMetric)
-          Duration.fromNanos(nanos)
-        }
-        perfLog = perfLog + (currentSize → toUpdate)
-      }
-      Some(
-        if (dispatchWaitWhenFullyUtilized.isDefined && Random.nextDouble() < explorationRatio)
+    case OptimizeOrExplore ⇒
+      val action = {
+        if (Random.nextDouble() < explorationRatio)
           explore(currentSize)
         else
           optimize(currentSize)
-      )
-    }
-
-  }
+      }
+      processor ! action
 
-  private def downsize: Option[ScaleTo] = {
-    underUtilizationStreak.flatMap { streak ⇒
-      if (streak.start.isBefore(LocalDateTime.now.minus(downsizeAfterUnderUtilization)))
-        Some(ScaleTo((streak.highestUtilization * downsizeRatio).toInt, Some("downsizing")))
-      else
-        None
-    }
+    case qs: QueryStatus ⇒
+      qs.reply(AutoScalingStatus(poolSize = Some(currentSize), performanceLog = perfLog))
   }
 
   private def optimize(currentSize: PoolSize): ScaleTo = {
```

Comment:
To give you some context, the method is called `optimize` but it's just a step in the whole optimization algorithm and it's a rather simple step - find an adjacent local optimal point and move towards it. Right now it's tested through the AutoScaling actor, it would be hard to test a method in an actor, you would have to move it to another class and dependency inject it, but since the logic is actually pretty simple, I don't know if it justify the complexity.